src/interface.jl

```diff
@@ -184,168 +229,178 @@ Possible values are:
+"""
+@multi_agent_env_api get_state(env::AbstractEnv, player=get_current_player(env)) = env.state
 
-#####
-## Observation
-#####
+"""
+    get_current_player(env)
 
+Return the next player to take action.
+For [Extensive Form Games](https://en.wikipedia.org/wiki/Extensive-form_game),
+A *chance player* may be returned. (See also [`get_chance_player`](@ref))
+For [SIMULTANEOUS](@ref) environments, a *simultaneous player* may be returned.
+(See also [`get_simultaneouse_player`](@ref)).
 """
-    observe(env::AbstractEnv) = observe(env, get_current_player(env))
-    observe(::AbstractEnv, player)
+@env_api get_current_player(env::AbstractEnv) = DEFAULT_PLAYER
 
-Get an observation of the `env` from the perspective of an `player`.
+"""
+    get_chance_player(env)
 
-!!! note
-    This is a very deliberate decision to adopt the duck-typing here
-    to describe an observation from an environment.
-    By default, we assume an observation is a NamedTuple,
-    which is the most common case.
-    But of course it can be of any type, as long as it implemented 
-    the necessay methods described in this section.
+Only valid for environments with a chance player.
 """
-@interface observe(env::AbstractEnv) = observe(env, get_current_player(env))
-@interface observe(::AbstractEnv, player)
+@env_api get_chance_player(env::AbstractEnv)
 
 """
-    get_legal_actions_mask(obs) -> Bool[]
+    get_simultaneouse_player(env)
 
-Only valid for observations of [`FULL_ACTION_SET`](@ref).
+Only valid for environments of [`SIMULTANEOUS`](@ref) style.
 """
-@interface get_legal_actions_mask(obs) = obs.legal_actions_mask
+@env_api get_simultaneouse_player(env)
 
 """
-    get_legal_actions(obs)
+    get_spectator_player(env)
 
-Only valid for observations of [`FULL_ACTION_SET`](@ref).
+Used in imperfect multi-agent environments.
 """
-@interface get_legal_actions(obs) = findall(get_legal_actions_mask(obs))
+@env_api get_spectator_player(env::AbstractEnv)
+
+@env_api get_players(env::AbstractEnv) = (DEFAULT_PLAYER,)
+
+@env_api get_num_players(env::AbstractEnv) = get_num_players(NumAgentStyle(env))
+
+get_num_players(::SingleAgent) = 1
+get_num_players(::MultiAgent{N}) where N = N
+
+"Reset the internal state of an environment"
+@env_api reset!(env::AbstractEnv)
+
+"Set the seed of internal rng"
+@env_api seed!(env::AbstractEnv, seed)
 
-get_legal_actions(
-    obs::NamedTuple{(:reward, :terminal, :state, :legal_actions, :legal_actions_mask)},
-) = obs.legal_actions
-get_legal_actions(obs::NamedTuple{(:reward, :terminal, :state, :legal_actions)}) =
-    obs.legal_actions
+"Get all actions in each ply"
+@multi_agent_env_api get_history(env::AbstractEnv, player=get_current_player(env))
 
 """
-    get_terminal(obs)
+    get_terminal(env, player=get_current_player(env))
 """
-@interface get_terminal(obs) = obs.terminal
+@multi_agent_env_api get_terminal(env::AbstractEnv, player=get_current_player(env)) = env.terminal
```

Comment:
automatically generate the method with  `player=get_current_player(env)` in @multi_agent_env_api